cs/cs/Serializer/VowpalWabbitMultiExampleSerializerCompiler.cs

```diff
@@ -0,0 +1,91 @@
+            var multiFeature = allFeatures.FirstOrDefault(fe => fe.Name == VowpalWabbitConstants.MultiProperty);
+            if (multiFeature == null)
+                return null;
+
+            // multi example path
+            // IEnumerable<> or Array
+            var adfType = InspectionHelper.GetEnumerableElementType(multiFeature.FeatureType);
+            if (adfType == null)
+                throw new ArgumentException("_multi property must be array or IEnumerable<>. Actual type: " + multiFeature.FeatureType);
+
+            var compilerType = typeof(VowpalWabbitMultiExampleSerializerCompilerImpl<,>).MakeGenericType(typeof(TExample), adfType);
+            return (IVowpalWabbitSerializerCompiler<TExample>)Activator.CreateInstance(compilerType, settings, allFeatures, multiFeature);
+        }
+
+        private sealed class VowpalWabbitMultiExampleSerializerCompilerImpl<TExample, TActionDependentFeature> : IVowpalWabbitSerializerCompiler<TExample>
+        {
+            private readonly VowpalWabbitSingleExampleSerializerCompiler<TExample> sharedSerializerCompiler;
+
+            private readonly VowpalWabbitSingleExampleSerializerCompiler<TActionDependentFeature> adfSerializerComputer;
+
+            private readonly Func<TExample, IEnumerable<TActionDependentFeature>> adfAccessor;
+
+            public VowpalWabbitMultiExampleSerializerCompilerImpl(VowpalWabbitSettings settings, List<FeatureExpression> allFeatures, FeatureExpression multiFeature)
+            {
+                Contract.Requires(settings != null);
+                Contract.Requires(allFeatures != null);
+                Contract.Requires(multiFeature != null);
+
+                var nonMultiFeatures = allFeatures.Where(fe => fe != multiFeature).ToList();
+
+                this.sharedSerializerCompiler = nonMultiFeatures.Count == 0 ? null :
+                    new VowpalWabbitSingleExampleSerializerCompiler<TExample>(
+                        nonMultiFeatures,
+                        settings == null ? null : settings.CustomFeaturizer,
+                        !settings.EnableStringExampleGeneration);
+
+                this.adfSerializerComputer = new VowpalWabbitSingleExampleSerializerCompiler<TActionDependentFeature>(
+                    AnnotationJsonInspector.ExtractFeatures(typeof(TActionDependentFeature)),
+                    settings == null ? null : settings.CustomFeaturizer,
+                    !settings.EnableStringExampleGeneration);
+
+                var exampleParameter = Expression.Parameter(typeof(TExample), "example");
+
+                // CODE condition1 && condition2 && condition3 ...
+                var condition = multiFeature.ValueValidExpressionFactories
+                    .Skip(1)
+                    .Aggregate(
+                        multiFeature.ValueValidExpressionFactories.First()(exampleParameter),
+                        (cond, factory) => Expression.AndAlso(cond, factory(exampleParameter)));
```

Comment:
It’s checking the no object along the access path is null.

Imagine

Example ex;
Ex.namespace1.feature1

If (ex.namespace1 != null)



In the general serializer the access path can be arbitrary long.